### PR TITLE
docs: ADR for text editor decisions May 2022

### PR DESCRIPTION
**Pull request by bszabo**
Mostly how to use and extended tinyMCE

### CODE REVIEW

**Comment by bszabo** [comment]
Teasing this document into two or more documents is no problem, if we deem it best to treat the design decisions individually

File: docs/decisions/0002-text-editor-adr.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Status
+
+Accepted
+
+# Context
+
+We seek to hoist the lecacy text (HTML) text editor out of the monolith into a decoupled React application.
```
Comment (rayzhou-bit) [inline docs/decisions/0002-text-editor-adr.md:7]: nit: spelling: legacy

File: docs/decisions/0002-text-editor-adr.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+
+
+# Decision set (each decision small)
+
+1. Rely on editor-wide Blockstore networking utilities for loading from, and saving to, XBlock content
+2. Continue using the tinyMCE editor
+3. Wrap the tinyMCE editor so as to be able to handle it as a vanilla React component
+4. Use a mix of configuration and tinyMCE API calls to customize the editor as needed 
+    * (e.g., offer image upload and image gallery selection)
```
[the remark this answers is not in the document]
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:19]: It's a good call. I'll add the text early on, in the section that spells out our goals

**Review by kenclary — APPROVED** [review]

File: docs/decisions/0002-text-editor-adr.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
+# Status
+
+Accepted
+
+# Context
+
+We seek to hoist the legacy text (HTML) text editor out of the monolith into a decoupled React application
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:7]: Note here that it is for the text xblock specifically. This is not a generalized text editor component.
```suggestion
We seek to hoist the legacy text (HTML) text xblock editor out of the monolith into a decoupled React application
```

File: docs/decisions/0002-text-editor-adr.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+without reducing editor capabilities in the process.
+
+The legacy editor delegates the bulk of its operations to the tinyMCE editor
+
+
+
+# Decision set (each decision small)
+
+1. Rely on editor-wide Blockstore networking utilities for loading from, and saving to, XBlock content
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:16]: Important note. **Blockstore** is a system for storing versioned, reusable educational content for Open edX. Blockstore, however, is presently only used as the storage method for V1 and V2 content libraries. Presently We use the [modulestore](https://github.com/openedx/edx-platform/blob/master/common/lib/xmodule/xmodule/modulestore/docs/overview.rst#overview-of-the-edx-modulestores) for our backend storage needs, but all of this is superflous-- what we need to tell the reader here is that we use the CMS' **xblock rest api**. Blockstore isn't being used at all, and doesn't point readers in the direction of the abstraction which provides information that is relevant to them.

File: docs/decisions/0002-text-editor-adr.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+
+Much of the editor configuration is actually defined in `TextEditor/pluginConfig.js`, and the editorHook mostly draws from this configuration to populate the editor."
+
+## Image upload/gallery selection
+The tinyMCE editor is extended to offer an image upload/select button. When the end-user pushes this button, the editor's window is occluded by a modal dialog capable of uploading an image. The end-user can also interact with this modal window to toggle between uploading a new image, or selecting from among previously uploaded images (aka, the image gallery), to save, or to cancel out.
+
+On its initialization, the tinyMCE editor is provided with an icon and a callback to associate with the image upload button.
+
+(per Ben Warzeski)
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:71]: nit: might want to remove this before we merge, as this ties documentation to specific people, which seems like a great way to generate more work for ben :)

File: docs/decisions/0002-text-editor-adr.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+2. Continue using the tinyMCE editor
+3. Wrap the tinyMCE editor so as to be able to handle it as a vanilla React component
+4. Use a mix of configuration and tinyMCE API calls to customize the editor as needed 
+    * (e.g., offer image upload and image gallery selection)
+
+
+
+# Consequences
+
+## Rely on editor-wide BlockStore networking decision
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:26]: This section, although it adroitly captures the request layer and the general network abstraction, might be best served as its own adr, as none of the information you captured is unique to the text editor.

File: docs/decisions/0002-text-editor-adr.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+The legacy editor delegates the bulk of its operations to the tinyMCE editor
+
+
+
+# Decision set (each decision small)
+
+1. Rely on editor-wide Blockstore networking utilities for loading from, and saving to, XBlock content
+2. Continue using the tinyMCE editor
+3. Wrap the tinyMCE editor so as to be able to handle it as a vanilla React component
+4. Use a mix of configuration and tinyMCE API calls to customize the editor as needed
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:19]: We haven't defined anywhere what is meant by tinyMCE API. Readers could be confused by this term. A worthy defintion is perhaps the api tinyMCE provides to perform operations on the editor object and its state.

**Review by connorhaugh — CHANGES_REQUESTED** [review]
Looks like you have done a nice job capturing what you have learned. I have added mostly some comments around terms, as well as a question on whether or not Decision 1. is something we want to document per editor, or for all V2 editors.

File: docs/decisions/0002-text-editor-adr.md (hunk 4 above)
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:16]: so, maybe "Rely on editor-wide networking utilities for loading from, and saving to, XBlock content"? (One word, Blockstore, deleted)

File: docs/decisions/0002-text-editor-adr.md (hunk 5 above)
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:71]: I'll mark Ben as a co-author for subsequent commits on this ADR

File: docs/decisions/0002-text-editor-adr.md (hunk 4 above)
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:16]: Per our out-of-band discussions, discussion on XBlock network support will be in a separate ADR with cross-editor scope.

File: docs/decisions/0002-text-editor-adr.md (hunk 7 above)
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:19]: TinyMCE had been mentioned earlier, in the Context section. But it hadn't been defined. I included a link to TinyMCE documentation, as well as the tag line used by TinyMCE to define itself.

File: docs/decisions/0002-text-editor-adr.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+1. Rely on cross-editor XBlock network access support, per "Network Request Layer" ADR in this repo
+2. Continue using the tinyMCE editor
+3. Wrap the tinyMCE editor so as to be able to handle it as a vanilla React component
+4. Use a mix of configuration and tinyMCE API calls to customize the editor as needed 
+    * (e.g., offer image upload and image gallery selection)
+    
+# Consequences
+
+## Rely on editor-wide BlockStore networking decision
```
Comment (connorhaugh) [inline docs/decisions/0002-text-editor-adr.md:24]: nit: Just a note here to also remove any reference to blockstore

**Review by connorhaugh — APPROVED** [review]
Just one semantic nit for you to fix, but this looks good!

File: docs/decisions/0002-text-editor-adr.md (hunk 8 above)
Comment (bszabo) [reply docs/decisions/0002-text-editor-adr.md:24]: I meant to say XBlock

**Review by jristau1984 — APPROVED** [review]

**Pull request merged**